File: 2025/2799_CountCompleteSubarraysinanArray.py
"""
You are given an array nums consisting of positive integers.

We call a subarray of an array complete if the following condition is satisfied:

The number of distinct elements in the subarray is equal to the number of distinct elements in the whole array.
Return the number of complete subarrays.

A subarray is a contiguous non-empty part of an array.



Example 1:

Input: nums = [1,3,1,2,2]
Output: 4
Explanation: The complete subarrays are the following: [1,3,1,2], [1,3,1,2,2], [3,1,2] and [3,1,2,2].
Example 2:

Input: nums = [5,5,5,5]
Output: 10
Explanation: The array consists only of the integer 5, so any subarray is complete. The number of subarrays that we can choose is 10.


Constraints:

1 <= nums.length <= 1000
1 <= nums[i] <= 2000
"""
from typing import List
from collections import Counter

class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:
        left, right, res = 0, 0, 0
        sliding_counter = Counter()
        element_num = len(set(nums))
        while (right < len(nums)):
            sliding_counter[nums[right]] += 1
            if len(sliding_counter) == element_num:
                while left < right and sliding_counter[nums[left]] > 1:
                    sliding_counter[nums[left]] -= 1
                    left += 1
                res += left + 1
            right += 1
        return res
    
sol = Solution()
nums = [1,3,1,2,2]
print(sol.countCompleteSubarrays(nums))
nums = [5,5,5,5]
print(sol.countCompleteSubarrays(nums))

File: CPP/2799_CountCompleteSubarraysinanArray.cpp
/*
 */
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
 public:
  int countCompleteSubarrays(const vector<int>& nums) {
    unordered_map<int, int> counter;
    int left = 0, right = 0, res = 0;
    unordered_set<int> num_set(nums.begin(), nums.end());
    int target_num = num_set.size();
    while (right < nums.size()) {
      counter[nums[right]] += 1;
      if (counter.size() == target_num) {
        while (left < right && counter[nums[left]] > 1) {
          counter[nums[left]] -= 1;
          left += 1;
        }
        res += left + 1;
      }
      right++;
    }
    return res;
  }
};

int main() {
  Solution sol;
  vector<int> nums = {1, 3, 1, 2, 2};
  cout << sol.countCompleteSubarrays(nums) << endl;
  nums = {5, 5, 5, 5};
  cout << sol.countCompleteSubarrays(nums) << endl;
}
